scripts/browserstack.py

```diff
@@ -430,15 +454,29 @@ def getSessionIdsForBuild(buildId):
             sessionIds.append(session["id"])
     return sessionIds
 
+def getSessionIdsAndDeviceForBuild(buildId):
+    sessionIds = []
+    buildStatus = get_build_status(buildId)
+    devices = buildStatus.json()["devices"]
+    for device in devices:
+        sessions_on_device = device["sessions"]
+        for session in sessions_on_device:
+            deviceIdentifier = f"{device['device']}-{device['os_version']}"
+            sessionIds.append({"session_id": session["id"], "device": deviceIdentifier})
+    return sessionIds
+
 
 def getFailedTestsForBuild(buildId):
-    sessionIds = getSessionIdsForBuild(buildId)
-    failedClasses = []
-    for sessionId in sessionIds:
-        failedClassesInSession = getFailedTestClassesForSession(buildId, sessionId)
+    sessionIdsAndDevices = getSessionIdsAndDeviceForBuild(buildId)
+    devicesWithFailedClasses = {}
+    for item in sessionIdsAndDevices:
+        failedClassesInSession = getFailedTestClassesForSession(buildId, item["session_id"])
         for failedClass in failedClassesInSession:
-            failedClasses.append(failedClass)
-    return classNamesToFullyQualifiedClassNames(failedClasses)
+            if item["device"] not in devicesWithFailedClasses:
+                devicesWithFailedClasses[item["device"]] = []
```

Comment:
not necessary, but if you want to simplify this, you can use a defaultdict. You import like so:
`from collections import defaultdict`
and then you'd initialize it like this:
`devicesWithFailedClasses = defaultdict(list)`

And you can then just safely call append on any key in the dict